package/MDAnalysis/lib/_cutil.pyx

```diff
@@ -47,47 +59,951 @@ ctypedef cset[int] intset
+                    _coords_add_vector64(coords[start:stop], dvecs[i])
+            else:
+                for i in range(nvec):
+                    mask = compound_masks[i]
+                    for j in range(mask.shape[0]):
+                        k = mask[j]
+                        # Don't use += here!
+                        coords[k, 0] = coords[k, 0] + dvecs[i, 0]
+                        coords[k, 1] = coords[k, 1] + dvecs[i, 1]
+                        coords[k, 2] = coords[k, 2] + dvecs[i, 2]
+        else:
+            for i in range(n):
+                coords[i, 0] = coords[i, 0] + dvecs[i, 0]
+                coords[i, 1] = coords[i, 1] + dvecs[i, 1]
+                coords[i, 2] = coords[i, 2] + dvecs[i, 2]
+
+
+def coords_center(float[:, :] coordinates not None, double[:] weights=None,
+                  np.intp_t[:] compound_indices=None, bint check_weights=False,
+                  bint return_compound_masks=False):
+    """Compute the center of a coordinate array.
+
+    If `weights` are supplied, the center will be computed as a weighted
+    average of the `coordinates`.
+
+    If `compound_indices` are supplied, the (weighted) centers per compound will
+    be computed.
+
+    If the weights (of a compound) sum up to zero, the weighted center (of that
+    compound) will be all ``nan`` (not a number). If `check_weights` is set to
+    ``True``, a :class:`ValueError` will be raised in that case.
+
+    Parameters
+    ----------
+    coordinates : numpy.ndarray
+        An array of dtype ``numpy.float32`` and shape ``(n, 3)`` containing the
+        coordinates to average.
+    weights : numpy.ndarray, optional
+        An array of dtype ``np.float64`` and shape ``(n,)`` containing the
+        weights for each coordinate.
+    compound_indices : numpy.ndarray, optional
+        An array of dtype ``numpy.intp`` and shape ``(n,)`` containing the
+        compound indices for each coordinate.
+    check_weights : bool, optional
+        If ``True``, raises a :class:`ValueError` if the weights (of any
+        compound) sum up to zero.
+    return_compound_masks : bool, optional
+        If ``True`` and `compound_indices` is not ``None``, an array of dtype
+        ``object`` containing index masks for each compound will be returned as
+        well. See :func:`unique_masks_int_1d` for details.
+
+    Returns
+    -------
+    numpy.ndarray
+        An array of dtype ``np.float64`` and shape ``(1, 3)`` or
+        ``(n_compounds, 3)`` containing the (weighted) center(s).
+
+    Raises
+    ------
+    ValueError
+        If the coordinates array has an invalid shape, or if the number of
+        coordinates, compound indices, or weights do not match.
+    ValueError
+        If `check_weights` is ``True`` and the weights (of any compound) sum up
+        to zero.
+
+
+    .. versionadded:: 0.20.0
+    """
+    cdef np.intp_t n = coordinates.shape[0]
+    cdef np.ndarray[np.float64_t, ndim=2] center
```

Comment:
old buffer interface

Reply:
This is there for a reason:
The `center` object is what the function returns in the end, and that should be a `numpy.ndarray`.
If I used a memoryview instead, I'd have to return `np.asarray(center)`. Since `center` is initialized with `np.zeros(...)`, which returns an `ndarray` anyway, this would mean throwing away this first `ndarray` object and creating a new one in the return statement, effectively killing performance.